### PR TITLE
[ll] Port dx12ll to dx12

### PR DESCRIPTION
**Pull request by msiglreith**
### Overview
* Port over (most/all) parts of the old dx12ll backend to the dx12 backend
* Remove dx12ll backend
* Implement first parts of hidden descriptor heaps for dx12 using a free-list allocator (deallocation missing)
* Allow descriptor sets to be mixed (srv/cbv/uav and samplers), but will split internally and requires up to 2x more root signature slots.

### TODO
* Pipeline barriers support is pretty basic atm, but it needs more love in general in `core`
* Renderpasses also only ported over and missing large chunks of functionality
* Descriptor pool deallocation
* Some functions haven't been implemented so far (including destruction calls), only trying to get `quad` running atm (and it works when manually specifiying the barriers),

### CODE REVIEW

File: examples/core/quad/main.rs [hunk 1]
```diff
@@ -173,15 +211,15 @@ fn main() {
     });
 
     pipeline_desc.attributes.push(pso::AttributeDesc {
-        location: 1,
+        location: 0,
```
Comment (kvark) [inline examples/core/quad/main.rs]: doesn't this make the existing Vk shaders incompatible?

File: src/backend/dx12/src/command.rs [hunk 2]
```diff
@@ -250,9 +351,43 @@ impl command::RawCommandBuffer<Backend> for CommandBuffer {
...
+            //       Resize while recording?
+            let mut heaps = [
+                sets[0].heap_srv_cbv_uav.as_mut() as *mut _,
+                sets[0].heap_samplers.as_mut() as *mut _
+            ];
+            self.raw.SetDescriptorHeaps(2, heaps.as_mut_ptr())
+        }
+
+        let mut table_id = 0;
+        for (i, table) in layout.tables.iter().enumerate() {
```
Comment (kvark) [inline src/backend/dx12/src/command.rs]: let's split this into 2 loops, one `for table in &layout.tables[..first_set]`, and another `for (set, _) in sets.iter_mut().zip(layout.tables[first_set..].iter())`

File: src/backend/dx12/src/device.rs [hunk 3]
```diff
@@ -1,82 +1,1144 @@
...
+                String::from_utf8_lossy(slice).into_owned()
+            };
+            return Err(d::ShaderError::CompilationFailed(message))
+        }
+
+        shader_map.insert(entry_point.into(), blob);
+        Ok(n::ShaderModule { shaders: shader_map })
+    }
+
+    pub fn create_descriptor_heap_impl(
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: nit: no need _impl suffix any more, given no descriptor heaps in the core trait

File: src/backend/dx12/src/device.rs [hunk 4]
```diff
@@ -1,82 +1,1144 @@
...
+            subpasses,
+        }
+    }
+
+    fn create_pipeline_layout(&mut self, sets: &[&n::DescriptorSetLayout]) -> n::PipelineLayout {
+        // Pipeline layouts are implemented as RootSignature for D3D12.
+        //
+        // Each descriptor set layout will be one table entry of the root signature.
+        // We have the additional restriction that SRV/CBV/UAV and samplers need to be
+        // separated, so each set layout will actually occupy 2 entries!
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: could we have a smarter logic that only splits a set layout if it has both samplers and non-samplers?

File: src/backend/dx12/src/device.rs [hunk 5]
```diff
@@ -1,82 +1,1144 @@
...
+
+            // Get color attachment formats from subpass
+            let (rtvs, num_rtvs) = {
+                let mut rtvs = [winapi::DXGI_FORMAT_UNKNOWN; 8];
+                let mut num_rtvs = 0;
+                for (mut rtv, target) in rtvs.iter_mut()
+                                             .zip(pass.color_attachments.iter())
+                {
+                    let format = subpass.main_pass.attachments[target.0].format;
+                    println!("{:?}", format);
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: debug print?

File: src/backend/dx12/src/device.rs [hunk 6]
```diff
@@ -1,82 +1,1144 @@
...
+
+        Ok(UnboundBuffer {
+            requirements,
+            stride,
+            usage,
+        })
+    }
+
+    fn get_buffer_requirements(&mut self, _buffer: &UnboundBuffer) -> Requirements {
         unimplemented!()
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: so lazy! :)

File: examples/core/quad/main.rs (hunk 1 above)
Comment (msiglreith) [reply examples/core/quad/main.rs]: No, it currently was wrongly implemented, switching uv and pos. Vulkan hasn't used the location but only the order they have been added to the attributes list. I think I included a fix for vulkan in this PR.

File: src/backend/dx12/src/device.rs (hunk 4 above)
Comment (msiglreith) [reply src/backend/dx12/src/device.rs]: It basically works this way, no empty tables. Comment doesn't reflect it properly (:

File: src/backend/dx12/src/device.rs (hunk 6 above)
Comment (msiglreith) [reply src/backend/dx12/src/device.rs]: :P

**Comment by torkleyy** [comment]
I'd recommend increasing timeout limit https://github.com/amethyst/amethyst/blob/develop/bors.toml#L20

**Comment by kvark** [comment]
Huge OSX backlog, as always :(

**Pull request merged**